streams/block/ciphers/kuznyechik/kuznyechik.cpp

```diff
@@ -0,0 +1,559 @@
+
+		b >>= 1;
+	}
+
+	return prod;
+}
+
+static uint8_t
+galois256_mul(uint8_t a, uint8_t b)
+{
+	int c;
+
+	if (a == 0 || b == 0)
+		return 0;
+
+	c = gf_log[a] + gf_log[b];
+
+	return gf_exp[c % 255];
+}
+
+static void
+galois_init_tables(uint8_t poly)
+{
+	int c;
+	int i;
+
+	c = 1;
+	for (i = 0; i < 256; i++) {
+		gf_log[c] = i;
+		gf_exp[i] = c;
+		c = galois256_mul_slow(c, GENERATING_POLY, poly);
+	}
+}
+
+static inline uint8_t
+l_func(const uint8_t *in, const struct kuzn_ctx *ctx)
+{
+	uint8_t prod;
+
+	prod  = galois256_mul(in[0x0], 148);
+	prod ^= galois256_mul(in[0x1], 32);
+	prod ^= galois256_mul(in[0x2], 133);
+	prod ^= galois256_mul(in[0x3], 16);
+	prod ^= galois256_mul(in[0x4], 194);
+	prod ^= galois256_mul(in[0x5], 192);
+	prod ^= in[0x6];
+	prod ^= galois256_mul(in[0x7], 251);
+	prod ^= in[0x8];
+	prod ^= galois256_mul(in[0x9], 192);
+	prod ^= galois256_mul(in[0xa], 194);
+	prod ^= galois256_mul(in[0xb], 16);
+	prod ^= galois256_mul(in[0xc], 133);
+	prod ^= galois256_mul(in[0xd], 32);
+	prod ^= galois256_mul(in[0xe], 148);
+	prod ^= in[0xf];
+
+	return prod;
+}
+
+//TODO: wipe magic constant
+static void
+R_func(uint8_t *out, const uint8_t *in, const struct kuzn_ctx *ctx)
+{
+	uint8_t tmp;
+	tmp = l_func(in, ctx);
+
+	memmove(out + 1, in, 0x0f);
+	out[0] = tmp;
+}
+
+static void
+Rinv_func(uint8_t *out, const uint8_t *in, const struct kuzn_ctx *ctx)
+{
+	uint8_t tmp;
+
+	tmp = in[0];
+	memmove(out, in + 1, 0x0f);
+	out[15] = tmp;
+
+	out[15] = l_func(out, ctx);
+}
+
+static void
+L_func(uint8_t *out, const uint8_t *in, const struct kuzn_ctx *ctx)
+{
+	int i;
+
+	for (i = 0; i < 0x10; i++)
+		R_func(out, in, ctx);
+}
+
+static void
+Linv_func(uint8_t *out, const uint8_t *in, const struct kuzn_ctx *ctx)
+{
+	int i;
+
+	for (i = 0; i < 0x10; i++)
+		Rinv_func(out, in, ctx);
+}
+
+static void
+F_func(uint8_t *out, const uint8_t *in, const uint8_t *c, const struct kuzn_ctx *ctx)
+{
+	uint8_t tmp[16];
+
+	LSX(tmp, in, c, ctx);
+
+	X_func(tmp, tmp, in + 0x10);
+
+	memcpy(out + 0x10, in, 0x10);
+	memcpy(out, tmp, 0x10);
+}
+
+static void
+kuzn_expand_key(struct kuzn_ctx *ctx)
+{
+	int i, j;
+	uint8_t *ptr;
+	uint8_t C[16];
+
+	ptr = ctx->keys + 32;
+
+	for (i = 0; i < 4; i++) {
+		memcpy(ptr, ptr - 32, 32);
+		for (j = 0; j < 8; j++) {
+			C_func(C, 8*i + j + 1, ctx);
+			F_func(ptr, ptr, C, ctx);
+		}
+		ptr += 32;
+	}
+}
+
+// API functions
+
+void
+kuzn_context_init(struct kuzn_ctx *ctx)
+{
+	assert(ctx != NULL);
+
+	memset(ctx, 0, sizeof(*ctx));
+
+	// p(x) = x**8 + x**7 + x**6 + x + 1
+	ctx->polynomial = KUZN_POLY;
+
+	galois_init_tables(ctx->polynomial);
+}
+
+void
+kuzn_encrypt(struct kuzn_ctx *ctx, const uint8_t *in, uint8_t *out, unsigned int rounds)
+{
+	uint8_t tmp[16];
+
+	assert(ctx != NULL && in != NULL && out != NULL);
+	for (int i = 0; i < rounds; ++i){
+        int j = i;
+		if (i == 0)
+			LSX(tmp, in, ctx->keys, ctx);
+		else {
+			if (i == 9) {
+				X_func(out, tmp, ctx->keys + (16 * j));
```

Comment:
Without this, all the computation will be stored and lost in `tmp` variable.

I would write it this way:

```c
        assert (rounds > 1 or rounds < 10);

	LSX(tmp, in, ctx->keys, ctx);		if (rounds == 1) goto copy_to_out;	//K1
	LSX(tmp, tmp, ctx->keys + 16, ctx);	if (rounds == 2) goto copy_to_out;	//K2
	LSX(tmp, tmp, ctx->keys + 16 * 2, ctx);	if (rounds == 3) goto copy_to_out;	//K3
	LSX(tmp, tmp, ctx->keys + 16 * 3, ctx);	if (rounds == 4) goto copy_to_out;	//K4
	LSX(tmp, tmp, ctx->keys + 16 * 4, ctx);	if (rounds == 5) goto copy_to_out;	//K5
	LSX(tmp, tmp, ctx->keys + 16 * 5, ctx);	if (rounds == 6) goto copy_to_out;	//K6
	LSX(tmp, tmp, ctx->keys + 16 * 6, ctx);	if (rounds == 7) goto copy_to_out;	//K7
	LSX(tmp, tmp, ctx->keys + 16 * 7, ctx);	if (rounds == 8) goto copy_to_out;	//K8
	LSX(tmp, tmp, ctx->keys + 16 * 8, ctx);	if (rounds == 9) goto copy_to_out;	//K9
	X_func(out, tmp, ctx->keys + 16 * 9);	if (rounds == 10) return;		//K10
copy_to_out:
        memcpy(out, tmp, 16);
```

The `goto`'s are solving issue with leaving temporal result in `tmp`.